### ISSUE TITLE
Support for Logical Types

### ISSUE THREAD

**Issue by lourencoteodoro** [aloneguid/parquet-dotnet#305]
### Issue description

I am looking into this library as an option to generate parquet files. I noticed the files generated by this library do not use Logical Types, we are only using Converted Types. If I use ParquetSharp I get Converted Types, but I would like to avoid using unmanaged code. Can you please clarify on why you do not support Logical Types and if there is any work being done in this area?

**Comment by aloneguid** [comment]
Thank you for asking the question. Currently, I find that ConvertedType is suitable for representing encoded types, and for backward compatibility, all readers are required to support it. While LogicalType can be useful for encoding decimals and Guids with richer information, I am unable to discern its broader usefulness. If it's not too much trouble, would you mind sharing your specific use case with me?

**Comment by aloneguid** [comment]
Actually i'm going to implement it anyway, stay tuned.

**Comment by lourencoteodoro** [comment]
I have found the same as you described. The main concern is if we implement a system that does not support logical types, but in the future logical types are added, that the file would be different (not a big deal, as it would still be compatible). I guess the main problem is just the impression we are not complying with the latest standard, even though I have not fully understood the motiviation behind this change in the standard.

**Comment by aloneguid** [comment]
Logical type support is pretty much ready, i'll try to release tomorrow ;)

**Comment by lourencoteodoro** [comment]
I have performed some tests and the files generated with 4.9 is throwing error's when trying to read with other tools. I will try to provide more details today.

**Comment by aloneguid** [comment]
I've tested with Spark who is happy as beans. How are you reading it?

**Comment by lourencoteodoro** [comment]
We are using these tools: 

https://pypi.org/project/parquet-tools/

**Comment by lourencoteodoro** [comment]
Let add all the details. I have created a very simple program like listed below:

``` csharp

        class Foo
        {
            public bool BooleanValue { get; set; }
            public sbyte SByteValue { get; set; }
            public byte ByteValue { get; set; }
            public short Int16Value { get; set; }
            public ushort UInt16Value { get; set; }
            public int Int32Value { get; set; }
            public uint UInt32Value { get; set; }
            public long Int64Value { get; set; }
            public ulong UInt64Value { get; set; }
            public float SingleValue { get; set; }
            public double DoubleValue { get; set; }
            public decimal DecimalValue { get; set; }
            public DateTime DateTimeValue { get; set; }
            public TimeSpan TimeSpanValue { get; set; }
            public string StringValue { get; set; }
        }
        static async Task LogicalTypesTest()
        {
            List<Foo> expectedList = new List<Foo>()
            {
                new Foo()
                {
                    BooleanValue = true,
                    SByteValue = sbyte.MinValue,
                    ByteValue = byte.MaxValue,
                    Int16Value = short.MinValue,
                    UInt16Value = ushort.MaxValue,
                    Int32Value = int.MinValue,
                    UInt32Value = uint.MaxValue,
                    Int64Value = long.MinValue,
                    UInt64Value = ulong.MaxValue,
                    SingleValue = float.Epsilon,
                    DoubleValue = double.Epsilon,
                    DecimalValue = decimal.One,
                    DateTimeValue = DateTime.UtcNow,
                    TimeSpanValue = TimeSpan.FromHours(1),
                    StringValue = "This is a string value. Here are some challenging Unicode characters: 𒈙🥔🇺🇸 Z̤͔ͧ̑̓ä͖̭̈̇lͮ̒ͫǧ̗͚̚o̙̔ͮ̇͐̇",
                },
            };

            using MemoryStream stream = new MemoryStream();
            await ParquetSerializer.SerializeAsync(expectedList, stream);

            stream.Position = 0;
            File.WriteAllBytes(@".\types.parquet", stream.ToArray());
        }
```

Then the file types.parquet gets generated. When I try to analyze this file using the python packages parquet-tools or parquet-cli they return errors.  

The following is the error for parquet-cli:
``` bash
 parq types.parquet -s

Traceback (most recent call last):
  File "/home/myhome/.local/bin/parq", line 8, in <module>
    sys.exit(main())
  File "/home/myhome/.local/lib/python3.6/site-packages/parq/main.py", line 41, in main
    pq_table = pq.read_table(cmd_args.file)
  File "/home/myhome/.local/lib/python3.6/site-packages/pyarrow/parquet.py", line 1915, in read_table
    coerce_int96_timestamp_unit=coerce_int96_timestamp_unit
  File "/home/myhome/.local/lib/python3.6/site-packages/pyarrow/parquet.py", line 1711, in __init__
    [fragment], schema=fragment.physical_schema,
  File "pyarrow/_dataset.pyx", line 978, in pyarrow._dataset.Fragment.physical_schema.__get__
  File "pyarrow/error.pxi", line 143, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 114, in pyarrow.lib.check_status
OSError: Could not open Parquet input source '<Buffer>': Metadata contains Thrift LogicalType that is not recognized
```
The following is the error for parquet-tools:

```
parquet-tools inspect types.parquet
Traceback (most recent call last):
  File "/home/myhome/.local/bin/parquet-tools", line 8, in <module>
    sys.exit(main())
  File "/home/myhome/.local/lib/python3.6/site-packages/parquet_tools/cli.py", line 26, in main
    args.handler(args)
  File "/home/myhome/.local/lib/python3.6/site-packages/parquet_tools/commands/inspect.py", line 45, in _cli
    filename=local_path,
  File "/home/myhome/.local/lib/python3.6/site-packages/parquet_tools/commands/inspect.py", line 52, in _execute
    pq_file: pq.ParquetFile = pq.ParquetFile(filename)
  File "/home/myhome/.local/lib/python3.6/site-packages/pyarrow/parquet.py", line 234, in __init__
    coerce_int96_timestamp_unit=coerce_int96_timestamp_unit
  File "pyarrow/_parquet.pyx", line 972, in pyarrow._parquet.ParquetReader.open
  File "pyarrow/error.pxi", line 114, in pyarrow.lib.check_status
OSError: Metadata contains Thrift LogicalType that is not recognized
```
If I downgrade to version 4.8 I do not get these errors.

**Comment by spanglerco** [comment]
I also see similar errors loading files created by 4.9.0 and later, both when I use [DataFusion](https://pypi.org/project/datafusion/) or [Dremio](https://dremio.com/) to read them:

```python
import datafusion

ctx = datafusion.SessionContext()
ctx.register_parquet("data", "data.parquet")
```

> Exception: DataFusion error: ParquetError(General("Could not parse metadata: bad data"))

I don't get errors using 4.8.1 of Parquet.NET or if I only use integer and string columns.

**Comment by aloneguid** [comment]
Thanks everyone, found the issue! Parquet.Net creates an empty `LogicalType` container if there is no corresponding locgical type mapping exists. Spark is fine with it, but `parquet-tools` from pypi isn't.

Before:

```
❯ parquet-tools show C:\dev\parquet-dotnet\src\Parquet.Test\bin\Debug\net7.0\temp.WriteQuestionableTypesTest.parquet
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\venv\shell\Scripts\parquet-tools.exe\__main__.py", line 7, in <module>
  File "C:\venv\shell\lib\site-packages\parquet_tools\cli.py", line 26, in main
    args.handler(args)
  File "C:\venv\shell\lib\site-packages\parquet_tools\commands\show.py", line 64, in _cli
    with get_datafame_from_objs(pfs, args.head) as df:
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "C:\venv\shell\lib\site-packages\parquet_tools\commands\utils.py", line 192, in get_datafame_from_objs
    df: Optional[pd.DataFrame] = stack.enter_context(pf.get_dataframe())
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\contextlib.py", line 492, in enter_context
    result = _cm_type.__enter__(cm)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "C:\venv\shell\lib\site-packages\parquet_tools\commands\utils.py", line 71, in get_dataframe
    yield pq.read_table(local_path).to_pandas()
  File "C:\venv\shell\lib\site-packages\pyarrow\parquet\core.py", line 2926, in read_table
    dataset = _ParquetDatasetV2(
  File "C:\venv\shell\lib\site-packages\pyarrow\parquet\core.py", line 2466, in __init__
    [fragment], schema=schema or fragment.physical_schema,
  File "pyarrow\_dataset.pyx", line 1004, in pyarrow._dataset.Fragment.physical_schema.__get__
  File "pyarrow\error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow\error.pxi", line 115, in pyarrow.lib.check_status
OSError: Could not open Parquet input source 'C:/dev/parquet-dotnet/src/Parquet.Test/bin/Debug/net7.0/temp.WriteQuestionableTypesTest.parquet': Metadata contains Thrift LogicalType that is not recognized
```

after:

```
❯ parquet-tools show C:\dev\parquet-dotnet\src\Parquet.Test\bin\Debug\net7.0\temp.WriteQuestionableTypesTest.parquet
+-------------+
|       qtype |
|-------------|
| 1.68243e+18 |
+-------------+
```

**Comment by aloneguid** [comment]
adding a fix in 4.9.2
